src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs

```diff
@@ -530,5 +530,57 @@ public static string Page(this IUrlHelper urlHelper, string pageName, object val
                 host: host,
                 fragment: fragment);
         }
+
+        /// <summary>
+        /// Generates an absolute URL for an action method, which contains the specified
+        /// <paramref name="action"/> name, <paramref name="controller"/> name, route <paramref name="values"/>,
+        /// <paramref name="protocol"/> to use, <paramref name="host"/> name, and <paramref name="fragment"/>.
+        /// Generates an absolute URL if the <paramref name="protocol"/> and <paramref name="host"/> are
+        /// non-<c>null</c>. See the remarks section for important security information.
+        /// </summary>
+        /// <param name="helper">The <see cref="IUrlHelper"/>.</param>
+        /// <param name="action">The name of the action method. When not defined, defaults to the current executing action.</param>
+        /// <param name="controller">The name of the controller. When not defined, defaults to the current executing controller.</param>
+        /// <param name="values">An object that contains route values.</param>
+        /// <param name="protocol">The protocol for the URL, such as "http" or "https".</param>
+        /// <param name="host">The host name for the URL.</param>
+        /// <param name="fragment">The fragment for the URL.</param>
+        /// <returns>The generated URL.</returns>
+        /// <remarks>
+        /// <para>
+        /// The value of <paramref name="host"/> should be a trusted value. Relying on the value of the current request
+        /// can allow untrusted input to influence the resulting URI unless the <c>Host</c> header has been validated.
+        /// See the deployment documentation for instructions on how to properly validate the <c>Host</c> header in
+        /// your deployment environment.
+        /// </para>
+        /// </remarks>
+        public static string ActionLink(
+            this IUrlHelper helper,
+            string action = null,
+            string controller = null,
+            object values = null,
+            string protocol = null,
+            string host = null,
+            string fragment = null)
```

Comment:
default values are the way of the future (5 years ago) I'm glad we're getting with the times

Reply:
Are we really now allowed to use this "modern" approach in public APIs? Was disallowed forever.

@Eilon ??

Reply:
Had a long chat with #weareallpranav. At this point, if we think this is truly the best API to create, I leave it to the area owners to make the call (@mkArtakMSFT and @rynowak )

Things to weigh:
* Having a million optional params isn't necessarily more usable than a million overloads (neither is quite the case here)
* Optional params are not self-descriptive about which combinations are meaningful and/or valid (if I set `host`, must I set `fragment` as well? If I set `values`, can I not set `controller`?)
* Is the API in practice and in common cases more usable one way or the other?

Reply:
I realise I left a glib remark, but take a look here: https://github.com/aspnet/AspNetCore/blob/master/src/Http/Routing/src/LinkGeneratorRouteValuesAddressExtensions.cs

After **attempting a lot of different API shapes** with this was the most usable and most reasonable design we found for this kind of API.

@dougbu - you can almost guess what I'm going to say "everything is case by case" - it's not that default parameter values *should be used everywhere* or *should never be used*, but sometimes if it's the right choice. If you're designing an API and you think that's the right choice, then it's worth at least considering.

Reply:
@mkArtakMSFT could you please clarify if by "proposed approach" you mean overloads or default values?

Cause the commit uses the default value approach, but existing APIs(non-absolute methods) use the overload approach.

In my opinion we should go with overloads to mirror the non-absolute counterparts and keep consistency.